drv/sidecar-seq-server/src/main.rs

```diff
@@ -24,177 +29,335 @@ use i2c_config::devices;
+            Tofino2Vid::V0P867 => set_vout(Volts(0.867)),
+            Tofino2Vid::V0P847 => set_vout(Volts(0.847)),
+            Tofino2Vid::V0P831 => set_vout(Volts(0.831)),
+            Tofino2Vid::V0P815 => set_vout(Volts(0.815)),
+            Tofino2Vid::V0P790 => set_vout(Volts(0.790)),
+            Tofino2Vid::V0P759 => set_vout(Volts(0.759)),
+        }
     }
 
-    fn led_off(&mut self) {
-        use drv_stm32xx_sys_api::*;
+    fn power_up(&mut self) -> Result<(), SeqError> {
+        ringbuf_entry!(Trace::InitiateTofinoPowerUp);
 
-        let sys = SYS.get_task_id();
-        let sys = Sys::from(sys);
-        sys.gpio_set_to(self.led, false).unwrap();
-        self.led_on = false;
+        // Initiate the power up sequence.
+        self.sequencer.set_enable(true)?;
+
+        // Wait for the VID to become valid, retrying if needed.
+        for i in 1..4 {
+            // Sleep first since there is a delay between the sequencer
+            // receiving the EN bit and the VID being valid.
+            hl::sleep_for(i * 25);
+
+            let maybe_vid = self.sequencer.vid().map_err(|e| {
+                if let FpgaError::InvalidValue = e {
+                    SeqError::InvalidTofinoVid
+                } else {
+                    SeqError::FpgaError
+                }
+            })?;
+
+            // Set Vout accordingy to the VID and acknowledge the change to the
+            // sequencer.
+            if let Some(vid) = maybe_vid {
+                self.apply_vid(vid)?;
+                self.sequencer.ack_vid()?;
+                ringbuf_entry!(Trace::TofinoVidAck);
+                return Ok(());
+            }
+        }
+
+        Err(SeqError::SequencerTimeout)
+    }
+
+    fn power_down(&mut self) -> Result<(), SeqError> {
+        ringbuf_entry!(Trace::InitiateTofinoPowerDown);
+        Ok(self
+            .sequencer
+            .set_enable(false)
+            .map_err(|_| SeqError::SequencerError)?)
     }
 
-    fn led_toggle(&mut self) {
-        if self.led_on {
-            self.led_off();
-        } else {
-            self.led_on();
+    fn handle_tick(&mut self) -> Result<(), SeqError> {
+        let state = self.sequencer.state()?;
+        let error = self.sequencer.error()?;
+
+        ringbuf_entry!(Trace::TofinoSequencerTick(self.policy, state, error));
+
+        match (self.policy, state, error) {
+            // Power down if the Tofino should be disabled.
+            (TofinoSequencerPolicy::Disabled, TofinoSeqState::InPowerUp, _) => {
+                self.power_down()
+            }
+            (TofinoSequencerPolicy::Disabled, TofinoSeqState::A0, _) => {
+                self.power_down()
+            }
+            // Power up
+            (
+                TofinoSequencerPolicy::LatchOffOnFault,
+                TofinoSeqState::A2,
+                TofinoSeqError::None,
+            ) => self.power_up(),
+            _ => Ok(()), // Do nothing by default.
```

Comment:
Why doesn't `RestartOnFault, A2, None` cause the system to power up?

For that matter, why doesn't `RestartOnFault, A2, Err(...)` cause the system to restart?